includes/acl/clip_reader/clip_reader.h

```diff
@@ -0,0 +1,450 @@
+#pragma once
+
+////////////////////////////////////////////////////////////////////////////////
+// The MIT License (MIT)
+//
+// Copyright (c) 2017 Nicholas Frechette & Animation Compression Library contributors
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "acl/core/memory.h"
+#include "acl/clip_reader/clip_reader_error.h"
+#include "acl/compression/animation_clip.h"
+#include "acl/sjson/sjson_parser.h"
+#include <functional>
+
+namespace acl
+{
+	class ClipReader
+	{
+	public:
+		ClipReader(Allocator& allocator, const char* const sjson_input, int input_length)
+			: m_allocator(allocator),
+			  m_parser(sjson_input, input_length),
+			  m_error()
+		{
+		}
+
+		~ClipReader()
+		{
+			m_allocator.deallocate(m_bone_names);
+		}
+
+		bool read()
+		{
+			if (m_read_already)
+			{
+				return false;
+			}
+
+			m_read_already = true;
+
+			if (read_version() &&
+				read_clip_header() &&
+				read_and_allocate_bones() &&
+				create_clip() &&
+				read_tracks() &&
+				nothing_follows())
+			{
+				return true;
+			}
+
+			if (m_clip != nullptr)
+			{
+				m_clip->~AnimationClip();
+				m_allocator.deallocate(m_clip);
+				m_clip = nullptr;
+			}
+
+			if (m_skeleton != nullptr)
+			{
+				m_skeleton->~RigidSkeleton();
+				m_allocator.deallocate(m_skeleton);
+				m_skeleton = nullptr;
+			}
+
+			return false;
+		}
+
+		AnimationClip* get_clip()
```

Comment:
when a getter returns a simple member or is a trivial short statement, I usually write them on a single line to avoid whitespace noise, personal preference